keylime/ima/ima_dm.py

```diff
@@ -0,0 +1,800 @@
+'''
+SPDX-License-Identifier: Apache-2.0
+Copyright 2022 Thore Sommer
+
+Parser and validator for device mapper IMA events
+- https://www.kernel.org/doc/html/v5.15/admin-guide/device-mapper/dm-ima.html
+'''
+import pickle
+import re
+
+from distutils.util import strtobool
+from itertools import chain
+from typing import Optional, List, Dict, Union
+from collections import ChainMap
+from dataclasses import dataclass
+from packaging import version
+
+import lark
+from lark import Lark, Transformer, v_args
+
+from keylime.failure import Component, Failure
+from keylime.common.algorithms import Hash
+from keylime.ima import ima_ast
+from keylime.ima.dm_grammar import DM_GRAMMAR
+
+
+class DeepChainMap(ChainMap):
+    """
+    Variant of ChainMap that handles updates for nested ChainMaps
+    Based on the example in https://docs.python.org/3/library/collections.html#collections.ChainMap
+    """
+
+    def __setitem__(self, key, value):
+        for mapping in self.maps:
+            if key in mapping:
+                mapping[key] = value
+                return
+        self.maps[0][key] = value
```

Comment:
nit for clarity: `super().__setitem(key, value)`

Reply:
This code is directly taken from the Python documentation, but I agree with you that this is more clear.